### PR TITLE
Disputes: Allow merchant to respond to inquiries from transaction detail page

### LINKED ISSUE #7193: Transaction Details > Dispute Details > Inquiry CTA buttons
Part of #6883

To the Transaction Details screen, we're adding buttons to the dispute details that the merchant can click to take action to resolve an inquiry (disputes with status `warning_needs_response`).

#### Additional Details
* Related issue for dispute actions: #6926
* Related issue for inquiry steps to resolve: #7245

### PR DESCRIPTION
**Pull request by brucealdridge**
Fixes #7193

#### Changes proposed in this Pull Request

Add handling of CTA and modal for when a dispute is of type Inquiry.

This PR mostly focuses on minor wording changes.

Inquiries have separate actions that can be undertaken as compared with a normal dispute. They can be challenged but not accepted. With Inquiries the `accept` action has been replaced with a `refund` action. 

When clicking the primary action `Submit Evidence` button, the user is taken through the normal flow to submit evidence for the dispute/inquiry. As with disputes, the secondary action button `Issue refund` pops up a modal with more information.

The primary action button inside the modal `View Order to Issue Refund`. It needs to be very clear to the merchant that they must issue a refund manually.  

An automatic refund ability may be added later, likely once #7248 is implemented.

#### Text changes

| | Dispute | Inquiry |
|--------|--------|--------|
| CTA Buttons | `Challenge Dispute` / `Accept Dispute` | `Submit Evidence` / `Issue Refund` |
| Modal Title | `Accept the dispute?` | `Issue a refund?` |
| Modal Content | Accepting the dispute marks it as Lost. The disputed amount will be returned to the cardholder, with a $15.00 USD dispute fee deducted from your account.<br /><br />Accepting the dispute is final and cannot be undone. | Issuing a refund will close the inquiry, returning the amount in question back to the cardholder. No additional fees apply.<br /><br />This action is final and cannot be undone.<br /><br />You will be taken to the order, where you must complete the refund process manually.. |
| Modal Buttons | `Cancel` / `Accept Dispute` | `Cancel` / `View Order to Issue Refund` | 

#### Screenshots

| Dispute | Inquiry |
|--------|--------|
| Screenshot 2023-09-27 at 1 46 44 PM | Screenshot 2023-09-27 at 1 46 51 PM |
| Screenshot 2023-09-27 at 1 47 05 PM | Screenshot 2023-09-27 at 1 46 59 PM |

#### Testing instructions

1. Enable feature flag by running `update_option( '_wcpay_feature_dispute_on_transaction_page', '1' );` in [WP Console](https://wordpress.org/plugins/wp-console/) or by modifying your database table `wp_options` directly.
2. Create a disputed  / inquiry transaction using the test card `4000000000001976` at checkout.
3. Visit the transaction details screen, it should show the `Submit Evidence` / `Issue Refund` buttons.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7298` or branch name `add/7193-dispute-cta-for-inquries` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=add/7193-dispute-cta-for-inquries) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: f33745d11810bf2e604968a226efdfd60d24591a
- Build time: 2023-10-04 03:22:49 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: client/payment-details/dispute-details/dispute-awaiting-response-details.tsx
Comment (Jinksi) [inline client/payment-details/dispute-details/dispute-awaiting-response-details.tsx]: Minor suggestion: use `isInquiry` rather than `! isInquiry` to improve readability and remain consistent with `isInquiry` checks below.

Also I suggest only capitalising the first word (`Submit evidence`) to remain consistent with `Challenge dispute`.
```suggestion
	const challengeButtonDefaultText = isInquiry( dispute )
		? __( 'Submit evidence', 'woocommerce-payments' )
		: __( 'Challenge dispute', 'woocommerce-payments' );

```

File: client/payment-details/dispute-details/dispute-awaiting-response-details.tsx
Comment (Jinksi) [inline client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:151]: Minor suggestion: explicitly define the props required, in this case we only need the `url`.
```suggestion
	orderDetails: Pick< OrderDetails, 'url' > | null;
```

File: client/payment-details/dispute-details/dispute-awaiting-response-details.tsx
Comment (Jinksi) [inline client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:227]: I find the nested `isInquiry` ternaries a bit hard to follow. This _might_ benefit from separate components, e.g. `DisputeActions` & `InquiryActions`.

This is not a major concern, however.

File: client/payment-details/dispute-details/dispute-awaiting-response-details.tsx
Comment (Jinksi) [reply client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:151]: Another option: ask only for the `orderUrl` in the component props.

File: client/payment-details/dispute-details/dispute-awaiting-response-details.tsx
Comment (Jinksi) [inline client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:339]: I think we should consider if we need a new tracks event here.

`wcpay_dispute_accept_click` doesn't seem like it is a suitable event for tracking when the merchant clicks `View Order to Issue Refund`.

File: client/payment-details/dispute-details/dispute-awaiting-response-details.tsx
Comment (Jinksi) [inline client/payment-details/dispute-details/dispute-awaiting-response-details.tsx]: I think the wording `This action is final and cannot be undone` is misleading here because Viewing the refund does not result in any irreversible action.

IMO we can remove this line for inquiries and let the order screen handle this warning (which it does, via a confirmation dialog).

File: client/payment-details/dispute-details/dispute-awaiting-response-details.tsx
Comment (Jinksi) [inline client/payment-details/dispute-details/dispute-awaiting-response-details.tsx]: Something doesn't look right about this icon. Is there an alternative we can use, e.g. `arrowRight`?

This is opinion only, so feel free to leave as-is.

Screenshot with `arrowRight`:

image

File: client/payment-details/dispute-details/dispute-awaiting-response-details.tsx
Comment (Jinksi) [inline client/payment-details/dispute-details/dispute-awaiting-response-details.tsx]: Minor suggestion: only capitalise the first word in the button label to match `Accept dispute`.
```suggestion
														'View order to issue refund',
```

**Review by Jinksi — CHANGES_REQUESTED** [review]
This tests well @brucealdridge 🙌

I think that there are some suggestions I've left that are worth addressing before merging:
* The `Action is final and cannot be undone` seems incorrect, since the action is viewing an order. I think this line should be removed within this modal for inquiries.
* Inconsistencies with button label capitalisation.

Other suggestions are minor/non-blocking.

Tests completed:
* Action buttons appear for inquiries awaiting a response ✅
* `Submit evidence` button takes me to the challenge form ✅
* `Issue refund` button displays a modal ✅
	* `Cancel` closes the modal ✅
	* `View order` button takes me to the correct order ✅
		* I tested the view order button in the (rare or impossible?) case of missing order URL. Error notice appears ✅
		* Subsequently refunding the order results in the inquiry being closed (in test-mode at least) ✅
			image
* Renders the `Continue with challenge` button label for a dispute with staged evidence ✅
	image


**Note:** I fixed broken tests in 7f24785 / 8a6166c and added tests for `warning_needs_response` in 24d86f5.

File: client/payment-details/dispute-details/dispute-awaiting-response-details.tsx
Comment (brucealdridge) [reply client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:339]: Is your concern here for the naming? Or the use of the same event?

I agree the naming is off, maybe something like `wcpay_dispute_modal_accept`

While it may not be super clear that the button says something different for Inquiries, we can note this and as the dispute status is passed through with this event it is distinguishable

File: client/payment-details/dispute-details/dispute-awaiting-response-details.tsx
Comment (brucealdridge) [reply client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:227]: I'm not sold on splitting the Dispute / Inquiry paths. But agree that it's not easily readable.

I'll have a go at breaking this up into something more manageable.

Comment (Jinksi) [reply client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:227]: I think this is quite subjective, so feel free to leave as-is!

File: client/payment-details/dispute-details/dispute-awaiting-response-details.tsx
Comment (Jinksi) [reply client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:339]: Reading the "event or props" part of the tracks cheat sheet ( PdjTHR-2FU-p2 ) helped to clarify my thoughts here:

> Sometimes you can use properties to distinguish between events in different parts of the UI, or with different parameters. The rule of thumb here is the event – what’s the user intention you’re tracking?
> - If the events track different things, use different events (not props).
> - If it’s the same user intention, with different options, use props for the options.

In this case, I think the user intention that we are tracking is different:
* Disputes: The user intent is accepting the dispute. `wcpay_dispute_accept_click` is fine here I think.
* Inquiries: The user intent is viewing the order page. This should have a distinct event, e.g. `wcpay_inquiry_view_order_click`

File: client/payment-details/dispute-details/dispute-awaiting-response-details.tsx
Comment (brucealdridge) [reply client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:227]: I tried to split it off but it was challenging given the `isModalOpen`, `setModal` and, `onModalClose` that had to be passed on.

I have opted instead for a function that returns the various properties rather than all the ternaries.

File: client/payment-details/dispute-details/dispute-awaiting-response-details.tsx
Comment (brucealdridge) [reply client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:339]: I've added a new event for both. I see them as simple UI, the user clicks `button` or the user clicks `modal button`. The actual acceptance or viewing of the order can be tracked separately.

I don't think it hurts at all to keep the tracking for inquiries/disputes separate here.

File: client/payment-details/dispute-details/dispute-awaiting-response-details.tsx
Comment (Jinksi) [reply client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:227]: **This PR works as intended, so I don't think my uncertainties below should hold up this PR being shipped.** This is a tricky subject that I struggle to find the correct answers for (see a very similar discussion in https://github.com/Automattic/woocommerce-payments/pull/7047#discussion_r1326775042).

I do think you've improved the readability of this code by the mapping of the labels etc  – this more clearly separates what is for inquiries and what is for disputes. 

However, I still find this entangled a bit and hard to follow. There are still some `isInquiry` ternaries outside of the modal, e.g. clicking `issue refund`/`accept` will do different things – inquiries probably wants to be a `Button href={orderUrl}` rather than `Button onClick={window.location.orderUrl}`.

I've explored another way to write this in #7339, the pseudocode TLDR is:

```tsx
const ExistingConciseMarkup = () => {
	const actionLabelsEtc = isInquiry ? {
		label: 'Inquiry Label',
	} : {
		label: 'Dispute Label',
	};
	
	// Markup is defined once, but logic is not as easy to follow.
	return (
		<Actions>
			{actionLabelsEtc.label}
		</Actions>
	);
}

// Repeat markup to make it clear what is for inquiries and what is for disputes
const SimpleButRepetitiveMarkup = () => {
	// Markup is defined twice, but logic is straightforward.
	return isInquiry ? (
		<Actions>Inquiry Label</Actions>
	) : (
		<Actions>Dispute Label</Actions>
	)
}
```

I'm unsure if this is _more_ or _less_ readable as a whole than the existing solution.

File: client/payment-details/dispute-details/dispute-awaiting-response-details.tsx
Comment (Jinksi) [reply client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:339]: Nice! I've updated the event names to what I think you intended (feel free to correct me if I'm mistaken). 5cb3447

```diff
-DISPUTE_INQUIRY_REFUND_CLICK: 'wcpay_dispute_modal_refund_click',
+DISPUTE_INQUIRY_REFUND_CLICK: 'wcpay_dispute_inquiry_refund_click',
-DISPUTE_INQUIRY_REFUND_MODAL_VIEW: 'wcpay_dispute_modal_refund_click',
+DISPUTE_INQUIRY_REFUND_MODAL_VIEW: 'wcpay_dispute_inquiry_refund_modal_view',
```

**Review by Jinksi — APPROVED** [review]
Looks good @brucealdridge!

Everything I tested before (https://github.com/Automattic/woocommerce-payments/pull/7298#pullrequestreview-1647684428) is working well ✅

I've left a comment discussing code readability that shouldn't prevent merging, approving ✅

File: client/payment-details/dispute-details/dispute-awaiting-response-details.tsx
Comment (haszari) [inline client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:188]: Splitting off the key bits of content to another helper function is making this harder to understand IMO. This component is tightly coupled to `getAcceptDisputeProps`. 

I don't fully understand the different conditions needed here (e.g. how many combinations are needed). It might be possible to simplify this code by having a single component with inline conditionals, or various components hard-coded for the common scenarios, and one conditional to choose which to render.

Comment (haszari) [reply client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:188]: Hunch… if we made a component for inquiries and a component for disputes, would that make this more readable and reduce the amount of data we need to pass down as props?

File: client/payment-details/dispute-details/dispute-awaiting-response-details.tsx
Comment (haszari) [inline client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:269]: Having to parameterise the button names and the tracks events is a signal that the UI is too generic. It's ok to implement a simple component with hard-coded content and single purpose.

File: client/payment-details/dispute-details/dispute-awaiting-response-details.tsx
Comment (haszari) [inline client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:296]: This is really generic and tightly coupled to `disputeAcceptAction` – to understand this code, requires reading whole file. How can we reduce that load?

Comment (haszari) [reply client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:296]: It's like an interface for generically rendering a modal. I don't think we need that capability, we can hard-code each modal we want as a self-contained component.

File: client/payment-details/dispute-details/dispute-awaiting-response-details.tsx
Comment (Jinksi) [reply client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:188]: > if we made a component for inquiries and a component for disputes

This was kind of the approach I explored in #7339.

I am tackling a similar problem with the inquiry steps to resolve PR (#7292) and I think separate components might help.

I don't think we should risk pushing back the project timeline for this, however. Can we ship this with 6.6 as-is and do a small tidy-up refactor once all of the code is in place?

**Comment by haszari** [comment]
I've added some comments – non blocking, I think there are ways we can simplify this code. However I don't fully understand it (need to spend a bit more time to do so), so I might be misguided.

Looking at the description, I think perhaps implementing a dedicated component for each modal might help a lot. 

@brucealdridge – can you clarify the title? I don't fully understand the changes – is this implementing wording changes for inquiries, on existing UI for disputes, and adding a modal flow to guide merchants to refund orders (in response to an inquiry)?

If so, I might summmarise that as `Allow merchant to respond to inquiries from transaction detail page`.

File: client/payment-details/dispute-details/dispute-awaiting-response-details.tsx
Comment (haszari) [reply client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:188]: Yeah totally, I don't want any of this to slow us down. We can always maintain, refactor, tidy in future :)

**Comment by brucealdridge** [comment]
I am merging despite failing one test which uses php7.3. The test fails as the latest WC requires php7.4.

File: client/payment-details/dispute-details/dispute-awaiting-response-details.tsx
Comment (haszari) [reply client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:188]: Reviewing this again as I register tracks events. I really want to refactor this, `AcceptDisputeProps` is an abomination!

File: client/payment-details/dispute-details/dispute-awaiting-response-details.tsx
Comment (haszari) [reply client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:296]: A side effect of this is that the tracks events both fire the same props, even though they are not relevant for both events. The tracks event name and the props should be nearby in the code (i.e. inline in the `recordEvent()` call) IMO.

File: client/payment-details/dispute-details/dispute-awaiting-response-details.tsx
Comment (Jinksi) [reply client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:188]: I started on this refactor in #7339, which may help or may not. I will close it if not.

Comment (haszari) [reply client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:188]: Awesome! @Jinksi Do you recall where you got to / if this was feeling like positive improvement? Great to have work-in-progress to build on, though I'm happy for you to keep moving on it.

Comment (Jinksi) [reply client/payment-details/dispute-details/dispute-awaiting-response-details.tsx:188]: I wrote up further in this thread that I was unsure if it was an improvement. I'll have to revisit this with fresh eyes and see.

https://github.com/Automattic/woocommerce-payments/pull/7298#discussion_r1343373487

**Pull request merged**
